### DISCUSSION TITLE
Possible support for Guide and Share buttons?

### DISCUSSION THREAD

**Discussion by aybe** [samuelgr/Xidi#22] [Ideas]
Any ideas on whether these two extra buttons could be supported as well by Xidi?

**Comment by samuelgr** [comment]
In theory this is possible but it would require some modifications to the XInput back-end that Xidi uses. XInput does not officially expose these 

Some undocumented functions apparently exist in the XInput library that could be leveraged, as in [here](https://gist.github.com/robindegen/9446175). This would at least enable the guide button but it comes with the risk of breaking if the XInput library ever changes going forward. I am not sure about the Share button though.

**Comment by aybe** [comment]
HID seems to be [one possible way](https://github.com/atar-axis/xpadneo/blob/67585b3a78202f01ea765deb1981533e4188c1cf/hid-xpadneo/src/hid-xpadneo.c#L849) to do so, however, there might be [an officially possible way](https://stackoverflow.com/a/55489270/361899) as the API isn't in the [not-supported list](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/desktop-to-uwp-supported-api), but that has to be tested. Either way it'll definitely be lower-level than stock XInput API so yeah, maybe not worth the trouble.

[This is also quite interesting](https://github.com/ViGEm/ViGEmBus/issues/40#issuecomment-619889180), apparently there's a kill switch about the triggers in the official API, I haven't been able to get it to work however, even after a reboot.

Right now your wrapper works beautifully for WinMM out of the box, so that'll do fine for now! 😀

**Comment by samuelgr** [comment]
Since we were discussing back-end changes: another option I was reading about is [XInputOnGameInput](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/reference/input/xinputongameinput/xinputongameinput_members) or possibly just [GameInput](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/input/porting/input-porting-xinput) itself. I do not think this would specifically allow support for the Guide and Share buttons though - while it does add [XInputSetStateEx](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/reference/input/xinputongameinput/functions/xinputsetstateex) for force feedback via the impulse triggers, there does not appear to be an equivalent to the undocumented XInputGetStateEx function.

One concern I have with the Guide and Share buttons being mappable in a game is that they are mapped to system-wide functions (at least the Guide button does by invoking the Xbox Game Bar - I don't have hardware with a Share button to know for sure). How would Xidi, or any other application that reads input from Guide and Share for that matter, interact with these system-wide functions? Users (or an app/Xidi) would presumably need to disable the system-level functions.

At least for the Guide button, users can [change a Game Bar setting](https://www.reddit.com/r/Steam/comments/jn9fu6/how_to_setup_xbox_controller_to_launch_big/) to turn off the system-wide function and suddenly DirectInput exposes the guide button as Button 11 without issue. Presumably the undocumented XInput function would do that too.

**Comment by aybe** [comment]
Never heard of that API before, but after [reading carefully](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/reference/input/xinputongameinput/functions/xinputgetstate#:~:text=Library%3A%20xgameruntime.lib-,Supported%20platforms%3A%20Xbox%20One%20family%20consoles%20and%20Xbox%20Series%20consoles,-See%20also), it's only for console...

Just tested both buttons, guide works as usual like you said, but share does absolutely nothing.

I've been searching fresh news about that:

https://www.windowscentral.com/xbox-series-x-s-share-button-customization-insiders#:~:text=Through%20the%20Xbox%20Accessories%20app,to%20a%20number%20of%20tasks.

https://www.theverge.com/2022/2/8/22923287/xbox-share-button-remap-controller-customization-feature

https://www.neowin.net/news/you-can-now-use-the-xbox-share-button-feature-on-windows-11-pcs-too

Looks like it'll be Windows 11 exclusive, because Microsoft you know... 🤣

**Comment by samuelgr** [comment]
It seems like the plan is to make GameInput [available to desktop apps](https://docs.microsoft.com/en-us/gaming/gdk/_content/gc/input/overviews/input-overview), which means Xidi could theoretically use it at some point.

Out of curiosity, even if it doesn't offer its intended system-wide functionality, does the Share button work via DirectInput? On Windows 10 or 11 if you search the Start menu for "Set up USB game controllers" and then click through to Properties, then press the Share button, does it cause any of the numbered buttons to light up on the Test tab?

**Comment by aybe** [comment]
Well that's two good news!

Totally forgot about that control panel and yes it has changed:

1. A
2. B
3. X
4. Y
5. Shoulder left
6. Shoulder right
7. Menu
8. Start
9. Thumb left
10. Thumb right
11. Guide
12. Share 😀
13. Nothing
14. Nothing
15. Nothing
16. Nothing

Hat as usual
Left thumbstick as usual
Right thumbstick is X/Y rotation but inverted on Y axis
Left trigger is +Z, Right trigger is -Z, shared as usual

So, we have the stuff we need and 4 unused slots, don't remember old version but on this one it only updates when the window is focused.

<img width="298" alt="rundll32_Kb7z7q1865" src="https://user-images.githubusercontent.com/1537591/153729736-af8eeb76-8968-4155-bc28-b628dfe192bc.png">

Just tried this on no$psx with WINMM:

```
[Mapper]
Type = Custom

[CustomMapper]
ButtonA = Button(11)
ButtonB = Button(12)
```

It works but the other way around, A sends Guide, B sends Share! 🤣

I tried `Button11 = Button(0)`:

```
---------------------------------------------
Xidi Log
---------------------------------------------
Version:   4.0.0
Form:      WinMM
Program:   C:\Temp\no$psx\NO$PSX.EXE
PID:       11492
---------------------------------------------
[02/12/2022 22:59:18] [E] Errors were encountered during configuration file reading.
[02/12/2022 22:59:18] [E]     C:\Temp\no$psx\Xidi.ini(7): Button11: Unrecognized configuration setting.
```

I tried `Button(0) = Button11`:

```
---------------------------------------------
Xidi Log
---------------------------------------------
Version:   4.0.0
Form:      WinMM
Program:   C:\Temp\no$psx\NO$PSX.EXE
PID:       21392
---------------------------------------------
[02/12/2022 23:00:20] [E] Errors were encountered during configuration file reading.
[02/12/2022 23:00:20] [E]     C:\Temp\no$psx\Xidi.ini(7): Unable to parse line.
```

**Comment by samuelgr** [comment]
Good news that Button 12 maps to Share. This confirms that DirectInput applications can certainly see it and might be a good indicator that there are some undocumented flags returned by `XInputGetState` and `XInputGetStateEx` that would see it as well.

By the way, that control panel is communicating with the controller without Xidi getting in the way at all. Xidi would need explicit support added for those new buttons. Currently all of its internal data structures completely exclude both Guide and Share.

```ini
; Guide button (could theoretically be mapped to anything, below might be what the built-in mapper XInputSharedTriggers would do)
ButtonGuide = Button(11)

; Share button (could theoretically be mapped to anything, below might be what the built-in mapper XInputSharedTriggers would do)
ButtonShare = Button(12)
```

The above is what the configuration might look like were Xidi to have support added for these two additional buttons. However, at the moment Xidi does not know about ButtonGuide or ButtonShare.

**Comment by aybe** [comment]
Which means that you'll need to mix both, XInput for separate triggers and DInput for guide/share buttons in that particular case, correct?

**Comment by samuelgr** [comment]
No, ideally everything would come from a single back-end like the undocumented XInput APIs or GameInput. I played around a bit with the undocumented XInput function `XInputGetStateEx` and it does work for the guide button, even though it is not officially supported.

**Reply by hifihedgehog** [reply]
The not-officially-supported language is unnecessarily scary. Mozilla has been safely supporting Guide button in their release Firefox codebase for many years with no fear or issues. Chrome-based browsers have likewise, presumably.

https://searchfox.org/firefox-main/source/dom/gamepad/windows/WindowsGamepad.cpp

I also managed to get Share working and stably for both a standard wired and Bluetooth wireless Xbox Series controllers. Stably was a bit trickier since it meant instituting a parallel background polling thread to reliably capture the Share button presses, which are not in the Xinput data. Let me know if this is of any use to anyone here!

https://github.com/hifihedgehog/Xidi/tree/2026-01-25.GuideAndShare
https://github.com/hifihedgehog/Xidi/commit/026a192035f4ac37e282f6eb591f51588dc191ac

**Comment by aybe** [comment]
Alright, I'm watching your repository to see what will be in your next release. 😀
